### ISSUE TITLE
Guidance on `Invoke-D365SDPInstall` usage 🚀

### ISSUE THREAD

**Issue by smholvoet** [d365collaborative/d365fo.tools#853]
Hi,

We've recenty switched from CHE dev environments to VHD based VM's which are being hosted on-prem.

I've ran into to the [Invoke-D365SDPInstall](https://github.com/d365collaborative/d365fo.tools/blob/master/d365fo.tools/functions/invoke-d365sdpinstall.ps1) cmdlet which looks like exactly what I need. It has a few examples, but I'm looking for the correct parameters to do the following on a VHD (stuff which was normally handled via LCS > Apply Update):

- Run a service update (e.g. update from 10.0.39 to 10.0.40)
- Install a package which was provided by an ISV (typically binaries) 

## Service Update

```PowerShell
$ServiceUpdate = "C:\Temp\FinanceAndOperations_10.0.1935.29_Application.zip"
Test-Path -Path $ServiceUpdate
Invoke-D365SDPInstall -Path $ServiceUpdate -QuickInstallAll
```

Output:

```
[15:11:26][Test-PathExists] The C:\Temp\FinanceAndOperations_10.0.1935.29_Application.zip\DefaultTopologyData.xml path wasn't found. Please ensure the path exists and you have enough permission to access the path.
[15:11:26][Test-PathExists] The C:\Temp\FinanceAndOperations_10.0.1935.29_Application.zip\AXUpdateInstaller.exe path wasn't found. Please ensure the path exists and you have enough permission to access the path.
WARNING: [15:11:26][Test-PathExists] Stopping because of missing paths.
```

EDIT: looks like the issue above was because I ran the command earlier which already extracted the ZIP. After pointing to the extracted folder instead of the .zip I now end up with:

```
VERBOSE: [15:35:53][Test-AssembliesLoaded] Total time spent inside the function was 00:00:00
VERBOSE: [15:35:59][Invoke-D365SDPInstall] Using QuickInstallAll mode
VERBOSE: [15:35:59][Invoke-Process] Output and Error streams will be redirected (silence mode)
VERBOSE: [15:35:59][Invoke-Process] Starting the AXUpdateInstaller.exe
VERBOSE: [15:35:59][Invoke-Process] Waiting for the AXUpdateInstaller.exe to complete
[15:36:00][Invoke-Process] Exit code from AXUpdateInstaller.exe indicated an error happened. Will output both standard stream and error stream.
[15:36:00][Invoke-Process] Standard output was: \r\n 
[15:36:00][Invoke-Process] Error output was: \r\n 
WARNING: [15:36:00][Invoke-Process] Stopping because of Exit Code. | Stopping because an Exit Code from AXUpdateInstaller.exe wasn't 0 (zero) like expected.
VERBOSE: [15:36:00][Invoke-D365SDPInstall] Total time spent inside the function was 00:00:06.0933456
```

This is also looks odd? 🧐

```PowerShell
PS C:\Temp\FinanceAndOperations_10.0.1935.29_Application> .\AXUpdateInstaller.exe list
List of Runbook ID on the system: 

List of services installed on the system: 

```

## ISV binaries

Does this require a different approach, compared to service updates?

Might be handy to include in the Wiki. Running version `0.7.14` should that matter.

**Comment by FH-Inway** [comment]
Thanks for the suggestion. `Invoke-D365SDPInstall` could indeed use some tutorial/documentation love, as it is one of the commonly used cmdlets of the module. Would you be interested in working with us to create such guidance? We will of course first help you to get up and running with your use cases.

For service updates, you want to use the `-Command RunAll` parameter as in [example 4](https://github.com/d365collaborative/d365fo.tools/wiki/Invoke-D365SDPInstall#example-4). For the ISV binaries, use `-DevInstall` as in [example 2](https://github.com/d365collaborative/d365fo.tools/wiki/Invoke-D365SDPInstall#example-2). The help for this parameter is not correct in my opinion. You always need to run the PowerShell as administrator to install a package.

I'm not sure why the help lists the `-QuickInstallAll` option as first example. Personally, I rarely use it and it also not officially documented by Microsoft. This should be considered a "know what you are doing" option in my opinion - and I would not know what I'm doing when using it.

Regarding the empty output by the `.\AXUpdateInstaller.exe list` command: Please update to version 0.7.15, which was released yesterday. It contains a fallback to address this issue. See #843 for more information on this.

**Comment by smholvoet** [comment]
Thank you Florian, updating to the latest version (`0.7.16`) did the trick 🎉

So service updates explicitely require the  [`-Command RunAll`](https://github.com/d365collaborative/d365fo.tools/blob/d8bac4670389b61aad6066f54c376c591328f3bf/d365fo.tools/functions/invoke-d365sdpinstall.ps1#L239) in order to run through all (what's in a name) steps, similar to what happens in the background if you trigger a service update through LCS.

I was also able to install the ISV binaries, but I'm still a bit confused as to why this requires `-DevInstall`? [Microsoft Docs](https://learn.microsoft.com/en-us/dynamics365/fin-ops-core/dev-itpro/deployment/install-deployable-package#install-an-application-aot-deployable-package-on-a-development-environment) as well as the inline comments [here](https://github.com/d365collaborative/d365fo.tools/blob/d8bac4670389b61aad6066f54c376c591328f3bf/d365fo.tools/functions/invoke-d365sdpinstall.ps1#L82C8-L82C159) mention the following:
> This command is to be used on Microsoft Hosted Tier1 development environment, where you don't have access to the administrator user account on the vm.

On a VHD I can simply run PowerShell as an admin user? So as you mentioned, this is a bit confusing.

On a sidenote and very much a personal opinion: the naming of this cmdlet, namely _Invoke-D365**SDP**Install_ is a bit weird in my opinion. It obviously refers to **S**oftware **D**eployable **P**ackage but it's easy to overlook this (very powerful) cmdlet when you're browsing through the contents of this PowerShell Module. I actually stumbled upon it via a community post even though I'm a long time user of this module. 

When I think of a "package" within a D365 F&O context, I think of deployable packages. Perhaps this was done to avoid any confusion with SqlPackage, but I'm sure most people would be able to make that distinction. A rename to `Invoke-D365PackageInstall` might be something to consider, up to you guys. The same would be needed for any other cmdlets which use the same abbrevation: `Get-D365SDPCleanUp`, `Set-D365SDPCleanUp` and `Get-D365SDPDetails`,

**Comment by FH-Inway** [comment]
Glad it's working for you.

The way I think about `-DevInstall` is that it is a subset of `-Command RunAll`. During `RunAll`, all the service models listed in the topology file get updated. `DevInstall` only updates the `AOSService` service model, which is usually all that is needed for ISV solutions.

As to why it supposedly does not require administrator privileges, I'm not sure. I have not tried it, but would expect it to fail when not running it as admin. There may have been a time where this was not the case, the parameter and its documentation was added in 2018. A lot has changed since then. This may have been a scenario when T2+ environments could still be accessed via RDP.

Regarding the sidenote: I hear you, when I started with d365fo.tools, I also had trouble finding cmdlets. And the naming of this particular one also raised my eyebrows :smile:
We could easily add aliases for the `SDP` cmdlets, though that would not help much with discoverability. The file names would still have the old name. To change that, the actual name would have to be changed and then the old name could be added as an alias. It is also weird that the cmdlet does not use `Install` as a [verb](https://learn.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-7.4#lifecycle-verbs:~:text=Start%2C%20Begin-,Install%20(is),-Places%20a%20resource). `Install-D365DeployablePackage` would be better in my opionion.

You may also be interested in `Get-Help about_deployable_packages`, which will give you [about_Deployable_Packages.help.txt](https://github.com/d365collaborative/d365fo.tools/blob/development/d365fo.tools/en-us/about_Deployable_Packages.help.txt). [Work with packages, resource label files, language and lables](https://github.com/d365collaborative/d365fo.tools/wiki/Work-with-packages,-resource---label-files,-language-and-lables) might also be of interest.

I'm closing this issue, as it seems to be resolved. But feel free to write further comments or open up a new issue or discussion.

**Comment by Splaxi** [comment]
Thanks for the feedback - if only I had a time machine, we could fix a lot of these things 😅

Back in the early days, we were working intensely on adding features and support the broader community, working bleeding edge on many things. During that point in time, we **might** (hehe) made some naming mistakes - which locks us today.

The reason for that, is that I'm a very firm believer in non-breaking changes. Next thing is that I don't want to keep to identical versions of a given feature alive - so that puts us in a tight spot in terms of evolving the module.

With that said - if people are up for the challenge to refactor / rename, and willing to put in the effort to unblock people that might hit road blocks based on the change, I'm up for the challenge. But doing it solo - I'm very focused on adding value and not polishing things for the sake of the sparkle/glitter.
